plugin-server/src/worker/ingestion/person-state.ts

```diff
@@ -450,57 +467,141 @@ export class PersonState {
-        //    `person.uuid` has already been used for events in ClickHouse, but they have no
-        //    corresponding rows in the `posthog_persondistinctid` or `posthog_person` tables
-        // 2. We need to assume the `distinct_id`/`person.uuid` have been used before (by
-        //    `$process_person_profile=false` events) and create an override row for this
-        //    `distinct_id` even though we're just now INSERT-ing it into Postgres/ClickHouse. We do
-        //    this by starting with `version=1`, as if we had just deleted the old user and were
-        //    updating the `distinct_id` row as part of the merge
-        const addDistinctIdVersion = 1
-
-        if (otherPerson && !mergeIntoPerson) {
-            await this.db.addDistinctId(otherPerson, mergeIntoDistinctId, addDistinctIdVersion)
-            return [otherPerson, Promise.resolve()]
-        } else if (!otherPerson && mergeIntoPerson) {
-            await this.db.addDistinctId(mergeIntoPerson, otherPersonDistinctId, addDistinctIdVersion)
-            return [mergeIntoPerson, Promise.resolve()]
+        // This means that at merge time, it's possible this `distinct_id` and its deterministically
+        // generated `person.uuid` has already been used for events in ClickHouse, but they have no
+        // corresponding rows in the `posthog_persondistinctid` or `posthog_person` tables.
+        //
+        // For this reason, $process_person_profile=false write to the `posthog_personlessdistinctid`
+        // table just to note that a given Distinct ID was used for "personless" mode. Then, during
+        // our merges transactions below, we do two things:
+        //   1. We check whether a row exists in `posthog_personlessdistinctid` for that Distinct ID,
+        //      if so, we need to write out `posthog_persondistinctid` rows with `version=1` so that
+        //      an override is created in ClickHouse which will associate the old "personless" events
+        //      with the Person UUID they were merged into.
+        //   2. We insert and/or update the `posthog_personlessdistinctid` ourselves, to mark that
+        //      the Distinct ID has been merged. This is important so that an event being processed
+        //      concurrently for that Distinct ID doesn't emit an event and _miss_ that a different
+        //      Person UUID needs to be used now. (See the `processPerson` code in `update` for more.)
+
+        if ((otherPerson && !mergeIntoPerson) || (!otherPerson && mergeIntoPerson)) {
+            // Only one of the two Distinct IDs points at an existing Person
+
+            const [existingPerson, distinctIdToAdd] = (() => {
+                if (otherPerson) {
+                    return [otherPerson!, mergeIntoDistinctId]
+                } else {
+                    return [mergeIntoPerson!, otherPersonDistinctId]
+                }
+            })()
```

Comment:
I did this to collapse the two different "one person exists, one doesn't" branches of the `if` into one shared code block. I like it, but open to opinions.